examples/magic_square.py

```diff
@@ -0,0 +1,474 @@
+(0, 2): ───H────┼@────────@───H───@───────┼────────────
+                ││                │       │
+(0, 3): ────────┼┼────────────H───@───H───M────────────
+                ││
+(1, 1): ───H────@┼────H───────────────H───M('bob')─────
+                 │                        │
+(1, 2): ───H─────@────H───────────────────M────────────
+               └──┘
+
+
+(row, col) = (2, 1)
+               ┌──┐
+(0, 0): ──────────────────────H───@───H───M('alice')───
+                                  │       │
+(0, 1): ───H────@─────────@───H───@───────┼────────────
+                │         │               │
+(0, 2): ───H────┼@────────@───H───@───────┼────────────
+                ││                │       │
+(0, 3): ────────┼┼────────────H───@───H───M────────────
+                ││
+(1, 1): ───H────@┼────H───────────────────M('bob')─────
+                 │                        │
+(1, 2): ───H─────@────H───────────────H───M────────────
+               └──┘
+
+
+(row, col) = (2, 2)
+               ┌──┐
+(0, 0): ──────────────────────────H───@───H───M('alice')───
+                                      │       │
+(0, 1): ───H────@─────────────@───H───@───────┼────────────
+                │             │               │
+(0, 2): ───H────┼@────────────@───H───@───────┼────────────
+                ││                    │       │
+(0, 3): ────────┼┼────────────────H───@───H───M────────────
+                ││
+(1, 0): ────────┼┼────────────────H───@───H───M('bob')─────
+                ││                    │       │
+(1, 1): ───H────@┼────H───H───@───H───@───────┼────────────
+                 │            │               │
+(1, 2): ───H─────@────H───────@───H───@───────┼────────────
+                                      │       │
+(1, 3): ──────────────────────────H───@───H───M────────────
+               └──┘
+
+
+Win rate for all 9 possible inputs, out of 10,000 repetitions each:
+[[1. 1. 1.]
+ [1. 1. 1.]
+ [1. 1. 1.]]
+
+"""
+
+
+import dataclasses
+
+import cirq
+import numpy as np
+
+
+def bell_pair_prep_circuit(q0: cirq.GridQubit, q1: cirq.GridQubit) -> cirq.Circuit:
+    """Prepare a Bell state between qubits q0 and q1.
+
+    Args:
+        q0: One qubit.
+        q1: The other qubit.
+
+    Returns:
+        A circuit creating a Bell state.
+    """
+    return cirq.Circuit(cirq.H.on_each(q0, q1), cirq.CZ(q0, q1), cirq.H(q1))
+
+
+def state_prep_circuit(
+    alice_data_qubits: tuple[cirq.GridQubit, cirq.GridQubit],
+    bob_data_qubits: tuple[cirq.GridQubit, cirq.GridQubit],
+) -> cirq.Circuit:
+    """Construct a circuit to produce the initial Bell pairs.
+
+    Args:
+        alice_data_qubits: Alice's data qubits.
+        bob_data_qubits: Bob's data qubits.
+
+    Returns:
+        A circuit preparing the Bell states.
+    """
+    pairs = ((alice_data_qubits[0], bob_data_qubits[0]), (alice_data_qubits[1], bob_data_qubits[1]))
+    return bell_pair_prep_circuit(*pairs[0]).zip(bell_pair_prep_circuit(*pairs[1]))
```

Comment:
Optional: Giving qubits short names and using destructuring bind (aka unpacking) makes it more readable:

```python
a1, a2 = alice_data_qubits
b1, b2 = bob_data_qubits
return bell_pair_prep_circuit(a1, b1).zip(bell_pair_prep_circuit(a2, b2))
```